gramps/gen/db/select_utils.py

```diff
@@ -0,0 +1,315 @@
+                    if row["handle"] not in handles_returned:
+                        handles_returned.append(row["handle"])
+                        yield row
+            return
+
+    # Otherwise, we have to scan the whole table:
+    iter_func = db._get_table_func(table.title(), "cursor_func")
+    for row in iter_func():
+        if where:
+            if match_where(row[1], where):
+                yield row[1]
+        else:
+            yield row[1]
+
+
+def sort_function(item, sort_by):
+    """
+    Given a list of jsonpaths, return a
+    tuple of associated values for sorting.
+    """
+    results = get_items(item, sort_by)
+    return tuple(results.values())
+
+
+def get_items(row, selections):
+    """
+    Given a list of jsonpaths, get the
+    items from the row.
+    """
+    results = {}
+    for json_path in selections:
+        if json_path == "$":
+            results = row
+        else:
+            match = match_json_path(json_path, row)
+            if match:
+                # Remove jsonpath syntax:
+                results[json_path[2:]] = match[0].value
+    return results
+
+
+def match_json_path(json_path, row):
+    """
+    Return the matching json_path item if the given
+    json_path matches.
+    """
+    if json_path not in PARSE_CACHE:
+        PARSE_CACHE[json_path] = jsonpath_ng.parse(json_path)
+    jsonpath_expr = PARSE_CACHE[json_path]
+    match = jsonpath_expr.find(row)
+    return match
+
+
+def eval_expr(expr, data):
+    """
+    Evaluate the expr. If the expr is a jsonpath,
+    then get the matching value, else all
+    other values are assumed to be self-evaluating.
+    """
+    if isinstance(expr, str) and expr.startswith("$"):
+        match = match_json_path(expr, data)
+        if match:
+            return match[0].value
+        else:
+            return None
+
+    return expr
+
+
+def match_where(data, where):
+    """
+    Assumes all comparisons are using jsonpath
+    ("$.gender", "=", value)
+    """
+    if len(where) == 3:
+        if where[1].lower() == "and":
+            for expr in [where[0], where[2]]:
+                result = match_where(data, expr)
+                if not result:
+                    return False
+            return True
+        elif where[1].lower() == "or":
+            for expr in [where[0], where[2]]:
+                result = match_where(data, expr)
+                if result:
+                    return True
+            return False
+        elif where[1].lower() in OPERATORS:
+            lhs = eval_expr(where[0], data)
+            op = where[1].lower()
+            rhs = eval_expr(where[2], data)
+            return compare_where(lhs, op, rhs)
+        else:
+            raise Exception("Operator %r is not supported" % where[1])
+    else:
+        raise Exception("Malformed where expression: %r" % where)
+
+
+def compare_where(lhs, op, rhs):
+    """
+    Evaluate the where expression.
+    """
+    if op == "=":
```

Comment:
would it be clearer to use match \ case instead of if \ elif?

Reply:
That's Python 3.10+.